File: leetcode/c++/0124-binary-tree-maximum-path-sum.cc
// https://leetcode.com/problems/binary-tree-maximum-path-sum/
// Hard

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int maxSoFar = 0;
    int maxPathSumRec(TreeNode* root) {
        int left = 0, right = 0;
        if (root->left) {
            left = maxPathSumRec(root->left);
        }
        if (root->right) {
            right = maxPathSumRec(root->right);
        }
        maxSoFar = max(max(root->val, maxSoFar), max(root->val + left, max(root->val + right, root->val + left + right)));
        return max (left+root->val, max(root->val, right+root->val));
    }
public:
    int maxPathSum(TreeNode* root) {
        maxSoFar = root->val;
        maxPathSumRec(root);
        return maxSoFar;
    }
};

File: leetcode/python/0124-binary-tree-maximum-path-sum.py
# https://leetcode.com/problems/binary-tree-maximum-path-sum/
# Hard

# Recursively call recurseMaxPathSum on both left and right subtrees, and max that to 0
# Keep track of the maxSoFar which is simply the max of itself and root.val + left + right
# In the recursive function, you return root.val + max(left, right) so that you
#   are only ever selecting one of the left or right subtrees
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxSoFar = float("-inf")

        def recurseMaxPathSum(root: Optional[TreeNode]) -> int:
            nonlocal maxSoFar

            if not root:
                return 0
            left, right = max(recurseMaxPathSum(root.left), 0), max(recurseMaxPathSum(root.right), 0)
            maxSoFar = max(maxSoFar, root.val + left + right)
            return root.val + max(left, right)
    
        recurseMaxPathSum(root)
        return maxSoFar